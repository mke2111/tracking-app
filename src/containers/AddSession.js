import React, { useEffect } from 'react';
import { useForm } from 'react-hook-form';
import { useDispatch, useSelector } from 'react-redux';
import { Redirect } from 'react-router-dom';
import {
  resetTaskData,
  setActiveTab,
  setDataError,
  setSessionTitle,
} from '../actions/index';
import autoLogin from '../api/autoLogin';
import { setSessionData } from '../api/setData';
import ErrorMessage from '../components/ErrorMessage';
import LoaderSpinner from '../components/LoaderSpinner';

const AddSession = () => {
  const token = localStorage.getItem('token');
  const { user, session } = useSelector(state => state);
  const dispatch = useDispatch();
  const { handleSubmit, register, errors } = useForm();

  const setTitle = e => {
    const input = e.target.value;
    dispatch(setSessionTitle(input));
  };

  useEffect(() => {
    if (token) { dispatch(setDataError(null)); }
    dispatch(autoLogin());
    dispatch(setActiveTab('Add Session'));
  }, []);

  const setSession = () => {
    dispatch(setSessionData());
    dispatch(resetTaskData());
  };
  const shouldComponentRender = () => {
    let isPending = false;
    if (user.pending === false || user.error !== null) {
      isPending = false;
    } else {
      isPending = true;
    }
    return isPending;
  };

  if (shouldComponentRender()) return <LoaderSpinner />;
  const errorText = `Error: ${user.error}`;

  return (
    <>
      {user.error && (
      <ErrorMessage message={errorText} />
      )}
      {!token && <Redirect to="/" />}
      {session.redirect && <Redirect to={`/sessionDetail/${session.session.id}`} />}
      {errors.title && (
        <ErrorMessage message={errors.title.message} />
      )}
      <form className="flex shadow rounded w-3/4 mx-auto my-8" onSubmit={handleSubmit(setSession)}>
        <input
          onChange={setTitle}
          name="title"
          type="text"
          className="m-3 outline-blue rounded-lg border-2 border-blue"
          placeholder="Add title"
          ref={register({ required: 'Field required' })}
        />
        <button className="w-36 rounded-md text-white bg-green-600" type="submit">Add</button>
      </form>
      <div className="w-4/6 mx-auto mt-36 text-center">
        <p>
          Add an activity that you want to track the time you spend doing. It could be Reading, Running, Coding, writing, Netflix N Chill, Baking, Surfing. <br/>YOU NAME IT!
        </p>

        <small className="text-green">
          Copyright 2023
        </small>
      </div>
    </>
  );
};

export default AddSession;
