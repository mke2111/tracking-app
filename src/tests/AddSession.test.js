import { render, screen, act } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../reducers/store';
import AddSession from '../containers/AddSession';

describe('Rendering component', () => {
  it('creates an AddSession component', () => {
    act(() => {
      render(
        <BrowserRouter>
          <Provider store={store}>
            <AddSession />
          </Provider>
        </BrowserRouter>,
      );
    });
    screen.getByText('Add');
  });
});

describe('Add Session', () => {
  it('renders correctly', () => {
    const snap = renderer.create(
      <BrowserRouter>
        <Provider store={store}>
          <AddSession />
        </Provider>
      </BrowserRouter>,
    ).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
