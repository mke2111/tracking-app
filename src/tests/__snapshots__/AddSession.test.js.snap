// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Session renders correctly 1`] = `
Array [
  <form
    className="flex shadow rounded w-3/4 mx-auto my-8"
    onSubmit={[Function]}
  >
    <input
      className="m-3 outline-blue rounded-lg border-2 border-blue"
      name="title"
      onChange={[Function]}
      placeholder="Add title"
      type="text"
    />
    <button
      className="w-36 rounded-md text-white bg-green-600"
      type="submit"
    >
      Add
    </button>
  </form>,
  <div
    className="w-4/6 mx-auto mt-36 text-center"
  >
    <p>
      Add an activity that you want to track the time you spend doing. It could be Reading, Running, Coding, writing, Netflix N Chill, Baking, Surfing.
       
      <br />
      YOU NAME IT!
    </p>
    <small
      className="text-green"
    >
      Copyright 2023
    </small>
  </div>,
]
`;
